core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -151,147 +205,159 @@ protected boolean resolveIndex() {
                     docWriteRequest.version(preVersions[j]);
                     docWriteRequest.versionType(preVersionTypes[j]);
                 }
-                throw (ElasticsearchException) e;
-            }
-            BulkItemRequest item = request.items()[requestIndex];
-            DocWriteRequest docWriteRequest = item.request();
-            if (isConflictException(e)) {
-                logger.trace((Supplier<?>) () -> new ParameterizedMessage("{} failed to execute bulk item ({}) {}",
-                        request.shardId(), docWriteRequest.opType().getLowercase(), request), e);
-            } else {
-                logger.debug((Supplier<?>) () -> new ParameterizedMessage("{} failed to execute bulk item ({}) {}",
-                        request.shardId(), docWriteRequest.opType().getLowercase(), request), e);
-            }
-            // if its a conflict failure, and we already executed the request on a primary (and we execute it
-            // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
-            // then just use the response we got from the successful execution
-            if (item.getPrimaryResponse() != null && isConflictException(e)) {
-                setResponse(item, item.getPrimaryResponse());
-            } else {
-                setResponse(item, new BulkItemResponse(item.id(), docWriteRequest.opType(),
-                    new BulkItemResponse.Failure(request.index(), docWriteRequest.type(), docWriteRequest.id(), e)));
+                throw e;
             }
+            // TODO: maybe this assert is too strict, we can still get environment failures while executing write operations
+            assert false : "unexpected exception: " + e.getMessage() + " class:" + e.getClass().getSimpleName();
```

Comment:
Now, we just throw an ElasticsearchException with a shardID as header instead of `OperationFailedEngineException` when we encounter a persistent engine failure. In bulk execution, we bubble it up (similar to retryOnPrimary exceptions)